File: set7/1/Main.java
import java.util.*;
import java.lang.*;
import java.io.*;

public class Main {
    public static void main(String[] args) {
        FastReader rd = new FastReader();

        int N = rd.nextInt();
        ArrayList<Tuple> V = new ArrayList<>();
        boolean []points = new boolean[2222222];
        final int base = 1010000;
        for(int i = 1; i <= N; ++i) {
            int x = rd.nextInt();
            int r = rd.nextInt();
            V.add(new Tuple(x - r, 1, i));
            V.add(new Tuple(x + r, -1, i));
            if(points[x - r + base] || points[x + r + base]) {
                System.out.print("NO");
                System.exit(0);
            }
            points[x - r + base] = points[x + r + base] = true;
        }
        Collections.sort(V);
        Stack<Integer> stack = new Stack<>();
        boolean ans = true;
        for(int i = 0; i < 2 * N; ++i) {
            Tuple value = V.get(i);
            if(value.b == 1) stack.push(value.c);
            else {
                if(stack.empty()) {
                    ans = false;
                    break;
                }
                else if(stack.peek() == value.c) stack.pop();
                else {
                    ans = false;
                    break;
                }
            }
        }
        if(ans) System.out.println("YES");
        else System.out.println("NO");
    }

    static public class Tuple implements Comparable<Tuple> {
        int a, b, c;
        public Tuple(int a, int b, int c) {
            this.a = a;
            this.b = b;
            this.c = c;
        }
        @Override
        public int compareTo(Tuple o) {
            if(this.a != o.a) return Integer.compare(this.a, o.a);
            return Integer.compare(this.b, o.b);
        }
    }

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while(st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                }
                catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() { return Integer.parseInt(next()); }
        long nextLong() { return Long.parseLong(next()); }
        double nextDouble() { return Double.parseDouble(next()); }
        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            }
            catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}

File: set7/1/main.cpp
#include<iostream>
#include<vector>
#include<tuple>
#include<stack>
#include<algorithm>

using namespace std;

vector<tuple<int, int, int>> v;
bool points[2020202];

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0);

    int N; cin >> N;
    const int base = 1010000;
    for(int i = 1; i <= N; i++) {
        int x, r; cin >> x >> r;
        v.emplace_back(x - r, 1, i);
        v.emplace_back(x + r, -1, i);
        if(points[x - r + base]) return cout << "NO", 0;
        if(points[x + r + base]) return cout << "NO", 0;
        points[x - r + base] = points[x + r + base] = 1;
    }
    sort(v.begin(), v.end());
    stack<int> st;
    bool ans = true;
    for(auto [x, t, idx]: v) {
        if(t == 1) st.emplace(idx);
        else {
            if(st.empty()) {
                ans = false;
                break;
            }
            else if(st.top() == idx) st.pop();
            else {
                ans = false;
                break;
            }
        }
    }

    if(ans) cout << "YES";
    else cout << "NO";

    return 0;
}